pymbolic/mapper/analysis.py

```diff
@@ -0,0 +1,59 @@
+__copyright__ = """Copyright (C) 2021 University of Illinois Board of Trustees"""
+
+__license__ = """
+Permission is hereby granted, free of charge, to any person obtaining a copy
+of this software and associated documentation files (the "Software"), to deal
+in the Software without restriction, including without limitation the rights
+to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+copies of the Software, and to permit persons to whom the Software is
+furnished to do so, subject to the following conditions:
+
+The above copyright notice and this permission notice shall be included in
+all copies or substantial portions of the Software.
+
+THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+THE SOFTWARE.
+"""
+
+
+from pymbolic.mapper import CachedWalkMapper
+
+
+__doc__ = """
+.. autofunction:: get_num_nodes
+"""
+
+
+# {{{ NodeCountMapper
+
+class NodeCountMapper(CachedWalkMapper):
+    """
+    Counts the number of nodes in an expression tree.
+
+    .. attribute:: count
+
+       The number of nodes.
+    """
+
+    def __init__(self) -> None:
+        super().__init__()
+        self.count = 0
```

Comment:
I think this could/should be done statelessly via a `CombineMapper`.

Reply:
I had thought about this during pytato. The problem here becomes the cache needs to store a "fake result" i.e. 0 for visited expressions.

Reply:
I've left it as a CachedWalkMapper for now.